File: tools/sysinfo.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# vim:fenc=utf-8
#
import os, sys, subprocess, gzip, logging, shutil, tarfile, os.path

#########################
##      globals
#########################
log = logging.getLogger("|")
log.setLevel(logging.DEBUG)
consoleHandler = logging.StreamHandler(sys.stdout)
consoleHandler.setFormatter(logging.Formatter(fmt='%(asctime)s %(levelname)-8s %(name)s - %(message)s',
    datefmt='%Y-%m-%d %H:%M:%S'))
log.addHandler(consoleHandler)

#########################
##      functions
#########################
def purgeFolder(folderPath):
    if os.path.exists(folderPath):
        log.info("cleaning %s" % folderPath)
        shutil.rmtree(folderPath)

def makeTar(output_filename, source_dir):
    with tarfile.open(output_filename, "w:xz") as tar:
        tar.add(source_dir, arcname=os.path.basename(source_dir))

#########################
##      main
#########################
log.info("adb wait-for-device ...")
subprocess.check_call("adb wait-for-device", shell = True)
subprocess.check_call("adb root", shell = True)
purgeFolder("sysinfo")
os.mkdir("sysinfo")
with open("sysinfo/0_prop", "wb") as f:
    f.write(subprocess.check_output("adb shell getprop", shell = True))
with open("sysinfo/1_partitions", "wb") as f:
    f.write(subprocess.check_output("adb shell cat /proc/partitions", shell = True))
    f.write(subprocess.check_output("adb shell ls -l /dev/block/by-name", shell = True))
with open("sysinfo/2_mount", "wb") as f:
    f.write(subprocess.check_output("adb shell mount", shell = True))
with open("sysinfo/3_kernel_cmdline", "wb") as f:
    f.write(bytes("[version]\n", "utf-8"))
    f.write(subprocess.check_output("adb shell cat /proc/version", shell = True))
    f.write(bytes("\n[cmdline]\n", "utf-8"))
    f.write(subprocess.check_output("adb shell cat /proc/cmdline", shell = True))
    f.write(bytes("\n[bootconfig]\n", "utf-8"))
    try:
        f.write(subprocess.check_output("adb shell cat /proc/bootconfig", shell = True))
    except subprocess.CalledProcessError as e:
        log.warning("can not read bootconfig")
        pass
    subprocess.check_call("adb pull /proc/config.gz", shell = True)
    with gzip.open("config.gz", "rb") as gz_file:
        file_content = gz_file.read()
        f.write(bytes("\n[defconfig]\n", "utf-8"))
        f.write(file_content)
    os.remove("config.gz")
subprocess.check_call("adb pull /proc/device-tree", cwd = "sysinfo", shell = True)
shutil.move("sysinfo/device-tree", "sysinfo/device_tree")
makeTar("sysinfo.py.tar.xz", "sysinfo")
shutil.rmtree("sysinfo")
log.info("sysinfo.py.tar.xz is ready")

File: lazybox/src/main/kotlin/cfig/lazybox/sysinfo/SysInfo.kt
package cfig.lazybox.sysinfo

import cfig.helper.Helper
import cfig.helper.Helper.Companion.check_call
import cfig.helper.ZipHelper
import org.slf4j.LoggerFactory
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.OutputStream
import java.nio.file.Files
import java.nio.file.Paths
import java.util.*
import kotlin.io.path.Path
import kotlin.io.path.deleteIfExists
import kotlin.io.path.writeText

class SysInfo {
    private fun runAndWrite(cmd: String, outStream: OutputStream, check: Boolean) {
        Helper.powerRun2(cmd, null).let {
            if (it[0] as Boolean) {
                outStream.write(it[1] as ByteArray)
            } else {
                if (check) {
                    log.warn(String(it[1] as ByteArray))
                    throw RuntimeException(String(it[2] as ByteArray))
                } else {
                    log.warn(String(it[1] as ByteArray))
                    log.warn(String(it[2] as ByteArray))
                }
            }
        }
    }

    fun makeTar(tarFile: String, srcDir: String, fmt: String) {
        val pyScript =
            """
import os, sys, subprocess, gzip, logging, shutil, tarfile, os.path
def makeTar(output_filename, source_dir):
   with tarfile.open(output_filename, "w:%s") as tar:
       tar.add(source_dir, arcname=os.path.basename(source_dir))
makeTar("%s", "%s")
""".trim()
        val tmp = Files.createTempFile(Paths.get("."), "xx.", ".yy")
        tmp.writeText(String.format(fmt, pyScript, tarFile, srcDir))
        ("python " + tmp.fileName).check_call()
        tmp.deleteIfExists()
    }

    private fun makeTar(tarFile: String, srcDir: String) {
        val pyScript =
            """
import os, sys, subprocess, gzip, logging, shutil, tarfile, os.path
def makeTar(output_filename, source_dir):
   with tarfile.open(output_filename, "w:xz") as tar:
       tar.add(source_dir, arcname=os.path.basename(source_dir))
makeTar("%s", "%s")
""".trim()
        val tmp = Files.createTempFile(Paths.get("."), "xx.", ".yy")
        tmp.writeText(String.format(pyScript, tarFile, srcDir))
        ("python " + tmp.fileName).check_call()
        tmp.deleteIfExists()
    }

    fun run() {
        "adb wait-for-device".check_call()
        "adb root".check_call()
        Path("sysinfo.tar").deleteIfExists()
        val prefix = "sysinfo"
        File("sysinfo").let {
            if (it.exists()) {
                log.info("purging directory sysinfo/ ...")
                it.deleteRecursively()
            }
        }
        File(prefix).mkdir()
        FileOutputStream("$prefix/0_prop").use {
            runAndWrite("adb shell getprop", it, true)
        }
        val theSlot: String = FileInputStream("$prefix/0_prop").use { inputStream ->
            Properties().apply {
                load(inputStream)
            }.getProperty("[ro.boot.slot_suffix]")?.let { slot ->
                when (slot) {
                    "[_a]" -> "_a"
                    "[_b]" -> "_b"
                    else -> {
                        log.warn("Unknown slot_suffix: $slot")
                        ""
                    }
                }
            } ?: run {
                log.warn("slot_suffix not found")
                ""
            }
        }

        FileOutputStream("$prefix/1_partitions").use { file ->
            runAndWrite("adb shell cat /proc/partitions", file, false) //HMOS
            runAndWrite("adb shell ls -l /dev/block/by-name", file, false)
        }

        FileOutputStream("$prefix/2_mount").use { file ->
            runAndWrite("adb shell mount", file, true)
        }

        FileOutputStream("$prefix/3_kernel_cmdline").use { file ->
            file.write("[version]\n".toByteArray())
            runAndWrite("adb shell cat /proc/version", file, true)
            file.write("\n[cmdline]\n".toByteArray())
            runAndWrite("adb shell cat /proc/cmdline", file, false)
            file.write("\n[bootconfig]\n".toByteArray())
            runAndWrite("adb shell cat /proc/bootconfig", file, false)
            // cpuinfo
            file.write("\n[cpuinfo]\n".toByteArray())
            runAndWrite("adb shell cat /proc/cpuinfo", file, false)
            // meminfo
            file.write("\n[meminfo]\n".toByteArray())
            runAndWrite("adb shell cat /proc/meminfo", file, false)
            // defconfig
            file.write("\n[defconfig]\n".toByteArray())
            "adb pull /proc/config.gz".check_call()
            ZipHelper.zcat("config.gz", "config")
            file.write(File("config").readBytes())
            File("config.gz").deleteOnExit()
            File("config").deleteOnExit()
        }

        FileOutputStream("$prefix/4_setting").use { fos ->
            fos.write("[global]\n".toByteArray())
            runAndWrite("adb shell settings list global", fos, false)
            fos.write("\n[system]\n".toByteArray())
            runAndWrite("adb shell settings list system", fos, false)
            fos.write("\n[secure]\n".toByteArray())
            runAndWrite("adb shell settings list secure", fos, false)
        }
        "adb pull /proc/device-tree".check_call(prefix)
        Files.move(Paths.get("$prefix/device-tree"), Paths.get("$prefix/device_tree"))

        if (theSlot.isNotBlank()) {
            "adb pull /dev/block/by-name/vbmeta$theSlot".check_call(prefix)
        }
        makeTar("sysinfo.tar.xz", "sysinfo")
        File("sysinfo").deleteRecursively()
    }

    companion object {
        private val log = LoggerFactory.getLogger(SysInfo::class.java)
    }
}